### DISCUSSION TITLE
Thoughts on separate DAGs to extract and process data?

### DISCUSSION THREAD

**Discussion by ldacey** [apache/airflow#29164] [General]
Historically, I have had DAGs which have tasks to download data from a source. Separate tasks within that DAG transform the data if needed and save it in Parquet format on cloud storage to be used for visualization and reporting.

One issue I have faced is that source data disappears over time outside of my control. APIs can archive data, SFTP servers delete old files, tables are truncated or archived.

When data-aware scheduling was released, I began to experiment with separating the extract tasks into a distinct DAG. I want to check with the community to see if it makes sense.

**extract-dag**
- Get the raw data to cloud storage (destination location is an Airflow Dataset which is used to trigger the next DAG)
- Trigger a processing DAG once complete

```
    extract_dataset = Dataset(params["raw_path"], extra={"version": 1})

    existing = list_files_task(retry_count=5, retry_minutes=60)
    retrieved = extract_file_task(file=existing, destination=extract_dataset)
    deleted = delete_file_task(retrieved)
    complete = EmptyOperator(task_id="trigger", outlets=[extract_dataset])

    existing >> retrieved >> [complete, deleted]
```

**process-dag**:
- Completely unaware of the source system of the data (SFTP, database, API, etc)
- Schedule is based on the `extract_dataset` above
- Lists files which were modified in the `extract_dataset` path since the schedule start time. This path is always on cloud storage regardless of the source system.
- Processes those files

```
    final_dataset = Dataset(params["analytics_path"], extra={"version": 1})

    existing = list_modified_blobs_task(path=params["raw_path"])
    transform = transform_file_task(source=existing, destination=staging_dataset)
    filters = create_filter_task(paths=transform)
    success = success_notification_task(files=existing)
    failure = failure_notification_task(files=existing)
    finalize = map_finalize_dataset_task(filter_or_paths=filters, destination=final_dataset)
    consolidate = map_consolidate_dataset_task(filter_or_paths=filters, destination=final_dataset)

    existing >> transform >> filters >> finalize >> [success, failure, consolidate]
```

It seems like a clean solution to have the data processing tasks be completely source-agnostic (before I would rely on XComs from the extract task to pass the list of files saved on cloud storage to be processed). I can also do a manaul DAG run trigger to re-process all files if needed, without worrying about clearing tasks.

One flaw is that it inflates my total DAGs and makes looking at logs more tedious if I want a full story. Maybe there will be better ways to view connected datasets in the future more cleanly though.

Any thoughts on this design pattern?

**Comment by WulkanHefajstos** [comment]
Hi!

We're using this type approach to build our pipelines  but we take it even a step further and we split the entire workflow into logical stages and create a separate DAG for each step (i.e. clean, normalize, etc.). This way our DAGs are pretty straightforward and relatively easy to maintain (usually they only consist of  a sensor(s) and a single operator responsible for processing the dataset) and that's most likely the biggest advantage of this approach. We've also noticed that it helps to ensure better quality / promote good standards while designing the pipeline, although that might be subjective feeling. Finally, it makes it relatively easy for other teams to "hook" into pipelines created by someone else (this way we can avoid processing the same dataset over and over again, just to create slightly different analytics).

To sync all of the DAGs we rely on common schedule interval and custom sensors / table partitions / external task markers / trigger dag run operator and we use folder structure, standardized names and tags to simplify the DAG discovery process.

As you pointed out, one of the reasons to organize our workflow like this was related to the fact that we have no control over the source data.

When it comes to the downsides - if the pipeline is not too complex, debugging is not that hard, you just need couple of extra mouse clicks when compared to DAG that includes the entire logic. Working with pipelines that consist of +20 DAGs is... well,annoying to put it mildly :D 

I didn't have a chance to use `Datasets` in production so i can't say anything meaningful about this - generally, it is a wonderful idea that could simplify our pipeline design, but atm it's not mature enough for most of our use cases. But i will definitely keep an eye on this 🙂

**Comment by potiuk** [comment]
> I didn't have a chance to use Datasets in production so i can't say anything meaningful about this - generally, it is a wonderful idea that could simplify our pipeline design, but atm it's not mature enough for most of our use cases. But i will definitely keep an eye on this slightly_smiling_face

This is precisely where Datasets should give you an order of magnitude better experience - especially whith our UI getting better with every release to make it easier to navigate. We even name this pattern "micropipelines" (micro-services anyone) and the bet a lot of the maintainers have is that this will become the "de-facto" standard on how DAGs/Pipelines are developed in Airlfow.

So I encourage both of you @WulkanHefajstos an @ldacey to take Datasets to a spin (in the latest version of Airflow)  if you are already convinced about the pattern and a) report back your experience and even better b) propose improvements if you find that some of the Dataset functionality could be still improved. I would love to hear back from you

**Comment by ldacey** [comment]
I have been using the Dataset feature across the board. It was very helpful for migrating from Azure to GCP because, as I mentioned in my original post, source data is not always available from the system.

1. Create a "producer" DAG which handles data extraction, but make the start_date something recent where the source data is present
2. Create a "consumer" DAG which is triggered from the success of "producer" and have it list files recently created in the Dataset.uri path (GCS bucket for me)
3. Manually trigger the "consumer" DAG to process historical files already on cloud storage (completely ignoring the "producer" DAG or source system)
